### PR TITLE
US5436: Edgex DevOps: New build automation for device-camera-go

### PR DESCRIPTION
**Pull request by vishwassp-hcl-intel**
Signed-off-by: Vishwas SP <vishwasx.s.premachandra@intel.com>

### CODE REVIEW

**Comment by vishwassp-hcl-intel** [comment]
Sandbox execution link: https://jenkins.edgexfoundry.org/sandbox/blue/organizations/jenkins/device-camera-go-verify-pipeline/detail/device-camera-go-verify-pipeline/5/pipeline

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        }
+
+        stage('SemVer Tag') {
+            when { expression { edgex.isReleaseStream() } }
+            steps {
+                unstash 'semver'
+                sh 'echo v${VERSION}'
+                edgeXSemver('tag')
+                edgeXInfraLFToolsSign(command: 'git-tag', version: 'v${VERSION}')
+                edgeXSemver('push')
```
Comment (ernestojeda) [inline Jenkinsfile]: please remove this line.

**Comment by vishwassp-hcl-intel** [comment]
> @vishwassp-hcl-intel One more change:
> 
> Can we change this line https://github.com/vishwassp-hcl-intel/device-camera-go/blob/US5436_change/Dockerfile#L1, to this:
> 
> ```
> ARG BASE=golang:1.12-alpine
> FROM ${BASE} AS builder
> ```
> 
> This will allow us to take advantage of the reusing the build image in the pipeline. While allowing the developers to build with `golang:1.12-alpine` locally.
> 
> Also, for Fuji, this version should be `1.1.0`: https://github.com/vishwassp-hcl-intel/device-camera-go/blob/US5436_change/Dockerfile#L2

OK, let me make this change

File: Jenkinsfile (hunk 1 above)
Comment (vishwassp-hcl-intel) [reply Jenkinsfile]: > please remove this line.

Changes are incorporated

**Comment by vishwassp-hcl-intel** [comment]
> @vishwassp-hcl-intel One more change:
> 
> Can we change this line https://github.com/vishwassp-hcl-intel/device-camera-go/blob/US5436_change/Dockerfile#L1, to this:
> 
> ```
> ARG BASE=golang:1.12-alpine
> FROM ${BASE} AS builder
> ```
> 
> This will allow us to take advantage of the reusing the build image in the pipeline. While allowing the developers to build with `golang:1.12-alpine` locally.
> 
> Also, for Fuji, this version should be `1.1.0`: https://github.com/vishwassp-hcl-intel/device-camera-go/blob/US5436_change/Dockerfile#L2

Changes are incorporated

**Comment by vishwassp-hcl-intel** [comment]
> Sandbox execution link: https://jenkins.edgexfoundry.org/sandbox/blue/organizations/jenkins/device-camera-go-verify-pipeline/detail/device-camera-go-verify-pipeline/5/pipeline

Updated after review comments incorporation:
https://jenkins.edgexfoundry.org/sandbox/blue/organizations/jenkins/device-camera-go-verify-pipeline/detail/device-camera-go-verify-pipeline/6/pipeline

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            // fan out
+            parallel {
+                stage('Build amd64') {
+                    stages {
+                        stage('Test') {
+                            steps {
+                                script {
+                                    // prep, ephemeral image is used as a caching layer
+                                    buildImage_amd64 = docker.build(
+                                        'image-base-build-amd64',
```
Comment (tobiasmo1) [inline Jenkinsfile]: Is there a risk of naming collision of this base image?

Dockerfile.build currently pulls in device-camera-go sources and dependencies. However, we are giving this base image a relatively generic name of `image-base-build-amd64`.

If the intent is to maintain a distinct base image/cache for each device service (DS), should we reflect the DS in the base image's name (e.g., '`image-base-build-amd64-dscamera`')?

Comment (ernestojeda) [reply Jenkinsfile]: @tobiasmo1 The `image-base-build-amd64` is just an ephemeral image used as a caching layer for all the gomod dependencies between the `go test` and `go build` stages. Once the build completes it is discarded.

**Comment by ernestojeda** [comment]
@vishwassp-hcl-intel The verify build failed due to missing settings file: `device-camera-go-codecov-token`. Please open a ticket with the LF to get that created.

**Pull request merged**